plugin.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// Secret holds the parameters of the SecretProviderClass CRD. Links the GCP
+// secret resource name to a path in the filesystem.
+type Secret struct {
+	// The resource name of the SecretVersion in the format projects/*/secrets/*/versions/*.
+	ResourceName string `json:"resourceName" yaml:"resourceName"`
+	// The filename where the contents of the secret are to be written.
+	FileName string `json:"fileName" yaml:"fileName"`
+}
+
+// StringArray holds the 'secrets' key of the SecretProviderClass parameters.
+// This is an array of yaml strings. Each string then must be parsed as a
+// 'Secret' struct.
+type StringArray struct {
+	Array []string `json:"array" yaml:"array"`
+}
+
+// plugin handles the secret volume mount events generated from the secrets CSI
+// driver. It fetches the secrets from the secretmanager API and writes them to
+// the filesystem based on the SecretProviderClass configuration.
+func plugin(ctx context.Context, client *secretmanager.Client, attributes, secrets, targetPath, permission string) error {
+	var attrib, secret map[string]string
+	var filePermission os.FileMode
+
+	// Everything in the "parameters" section of the SecretProviderClass.
+	if err := json.Unmarshal([]byte(attributes), &attrib); err != nil {
+		return fmt.Errorf("failed to unmarshal attributes: %v", err)
+	}
+
+	// The secrets here are the relevant CSI driver (k8s) secrets. See
+	// https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html
+	// Currently unused.
+	if err := json.Unmarshal([]byte(secrets), &secret); err != nil {
+		return fmt.Errorf("failed to unmarshal secrets: %v", err)
+	}
+
+	// Permissions to apply to all files.
+	if err := json.Unmarshal([]byte(permission), &filePermission); err != nil {
+		return fmt.Errorf("failed to unmarshal file permission: %v", err)
+	}
+
+	log.Printf("attributes: %v", attrib)
+	log.Printf("secrets: %v", secret)
+	log.Printf("filePermission: %v", filePermission)
+	log.Printf("targetPath: %v", targetPath)
+
+	var objects StringArray
+	if err := yaml.Unmarshal([]byte(attrib["secrets"]), &objects); err != nil {
```

Comment:
It's possible that `attrib` has no key "secrets". We should check and error before trying to unmarshal.

Reply:
done